pkg/meta/redis.go

```diff
@@ -2244,6 +2248,62 @@ func (r *redisMeta) doDeleteFileData_(inode Ino, length uint64, tracking string)
 	_ = r.rdb.ZRem(ctx, r.delfiles(), tracking)
 }
 
+func (r *redisMeta) doCleanupDelayedSlices(edge int64, limit int) (int, error) {
+	ctx := Background
+	stop := fmt.Errorf("reach limit")
+	var count int
+	err := r.hscan(ctx, r.delSlices(), func(keys []string) error {
+		for i := 0; i < len(keys); i += 2 {
+			key := keys[i]
+			ps := strings.Split(key, "_")
+			if len(ps) != 2 {
+				logger.Warnf("Invalid key %s", key)
+				continue
+			}
+			if ts, e := strconv.ParseInt(ps[1], 10, 64); e != nil {
+				logger.Warnf("Invalid key %s", key)
+				continue
+			} else if ts >= edge {
+				continue
+			}
+			buf := []byte(keys[i+1])
+			ss := r.decodeDelayedSlices(buf)
+			if ss == nil {
+				logger.Errorf("Invalid value for delSlices %s: %v", key, buf)
+				continue
+			}
+			rs := make([]*redis.IntCmd, len(ss))
+			if err := r.txn(ctx, func(tx *redis.Tx) error {
+				_, e := tx.Pipelined(ctx, func(pipe redis.Pipeliner) error {
+					for i, s := range ss {
+						rs[i] = pipe.HIncrBy(ctx, r.sliceRefs(), r.sliceKey(s.Chunkid, s.Size), -1)
+					}
+					pipe.HDel(ctx, r.delSlices(), key)
+					return nil
+				})
+				return e
+			}, r.delSlices()); err != nil {
```

Comment:
don't need to lock r.delSlices(), or we should use  tx.HGet(r.delSlices(), key) inside txn()

Reply:
There is no conflict, so no key is needed.

Reply:
At lease one key is required for Redis txn to use pessimistic lock